ocaml/typing/persistent_env.ml

```diff
@@ -344,23 +397,280 @@ let find_import ~allow_hidden penv ~check modname =
           add_import penv modname;
           acknowledge_import penv ~check modname psig
 
+let remember_global { globals; _ } global ~mentioned_by =
+  if Global_module.has_arguments global then
+    let global_name = Global_module.to_name global in
+    match Hashtbl.find globals global_name with
+    | exception Not_found ->
+        Hashtbl.add globals global_name
+          { gn_global = global; gn_mentioned_by = mentioned_by }
+    | { gn_global = old_global; gn_mentioned_by = first_mentioned_by } ->
+        if not (Global_module.equal old_global global) then
+          error (Inconsistent_global_name_resolution {
+              name = global_name;
+              old_global;
+              new_global = global;
+              first_mentioned_by;
+              now_mentioned_by = mentioned_by;
+            })
+
+let current_unit_is0 name ~allow_args =
+  match CU.get_current () with
+  | None -> false
+  | Some current ->
+      match CU.to_global_name current with
+      | Some { head; args } ->
+          (args = [] || allow_args)
+          && CU.Name.equal name (head |> CU.Name.of_string)
+      | None -> false
+
+let current_unit_is name =
+  current_unit_is0 name ~allow_args:false
+
+let current_unit_is_instance_of name =
+  current_unit_is0 name ~allow_args:true
+
 (* Enforce the subset rule: we can only refer to a module if that module's
    parameters are also our parameters. *)
-let check_for_unset_parameters penv modname import =
+let check_for_unset_parameters penv global =
+  (* A hidden argument specifies that the importing module should forward a
+     parameter to the imported module. Therefore it's the hidden arguments that
+     we need to check. *)
   List.iter
-    (fun param ->
-       if not (is_registered_parameter_import penv param) then
+    (fun ({ param = arg_name; value = arg_value } : Global_module.argument) ->
+       (* The _value_ is what we care about - the name lives in the imported
+          module's namespace, not ours *)
+       ignore arg_name;
+       let value_name = Global_module.to_name arg_value in
```

Comment:
I don't think there's any actual soundness concern here. The thing we want to know is whether a particular global module is registered as a parameter. If it's got arguments then the answer is definitely no, but it's a well-formed question.

Reply:
Yeah I wasn't thinking about soundness, but just trying to make invalid state non-representable. Because hidden args must come directly from the current's unit's parameter, so they should be represented as so. Currently we don't have `Parameter.t`, but we should make that fact clear by asserting it's just a `string`.

Reply:
Also, this should recurse deeply as well. I.e., the hidden argument of visible argument also need to be checked.

Reply:
> Yeah I wasn't thinking about soundness, but just trying to make invalid state non-representable. Because hidden args must come directly from the current's unit's parameter, so they should be represented as so. Currently we don't have `Parameter.t`, but we should make that fact clear by asserting it's just a `string`.

I don't think `string` would be an improvement here. At any rate, we can defer this to the `Parameter.t` PR.

Reply:
I agree, let's defer it to #3172 
WRT recursive hidden arguments - do you think we need to handle that?

Reply:
We don't need to, no (see the comment at the end of the function). Might be good to rename the function to make it clear it's only checking the top level of the global

Reply:
Rewrote the comments here (and added a comment at the call site) to hopefully clarify things, including why we don't need to recurse.